component/otelcol/internal/zapadapter/zapadapter.go

```diff
@@ -0,0 +1,240 @@
+// for specifying log levels.
+func New(l log.Logger) *zap.Logger {
+	return zap.New(&loggerCore{inner: l})
+}
+
+// loggerCore is a zap.Core implementation which forwards logs to a log.Logger
+// instance.
+type loggerCore struct {
+	inner log.Logger
+}
+
+var _ zapcore.Core = (*loggerCore)(nil)
+
+func (lc *loggerCore) Enabled(zapcore.Level) bool {
+	// An instance of log.Logger has no way of knowing if logs will be filtered
+	// out, so we always have to return true here.
+	return true
+}
+
+func (lc *loggerCore) With(ff []zapcore.Field) zapcore.Core {
+	enc := newFieldEncoder()
+	defer func() { _ = enc.Close() }()
+
+	for _, f := range ff {
+		f.AddTo(enc)
+	}
+
+	return &loggerCore{
+		inner: log.With(lc.inner, enc.fields...),
+	}
+}
+
+func (lc *loggerCore) Check(e zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	return ce.AddCore(e, lc)
+}
+
+func (lc *loggerCore) Write(e zapcore.Entry, ff []zapcore.Field) error {
+	enc := newFieldEncoder()
+	defer func() { _ = enc.Close() }()
+
+	enc.fields = append(enc.fields, "msg", e.Message)
+
+	for _, f := range ff {
+		f.AddTo(enc)
+	}
+
+	switch e.Level {
+	case zapcore.DebugLevel:
+		return level.Debug(lc.inner).Log(enc.fields...)
+	case zapcore.InfoLevel:
+		return level.Info(lc.inner).Log(enc.fields...)
+	case zapcore.WarnLevel:
+		return level.Warn(lc.inner).Log(enc.fields...)
+	case zapcore.ErrorLevel, zapcore.DPanicLevel, zapcore.PanicLevel, zapcore.FatalLevel:
+		// We ignore panics/fatals hwere because we really don't want components to
+		// be able to do that.
+		return level.Error(lc.inner).Log(enc.fields...)
+	default:
+		return lc.inner.Log(enc.fields...)
+	}
+}
+
+func (lc *loggerCore) Sync() error {
+	return nil
+}
+
+// fieldEncoder implements zapcore.ObjectEncoder. It enables converting a
+// zapcore.Field into a value which will be written as a github.com/go-kit/log
+// keypair.
+type fieldEncoder struct {
+	fields    []interface{}
+	namespace []string
+}
+
+var _ zapcore.ObjectEncoder = (*fieldEncoder)(nil)
+
+var encPool = sync.Pool{
+	New: func() any {
+		return &fieldEncoder{}
+	},
+}
+
+// newFieldEncoder creates a ready-to-use fieldEncoder. Call Close once the
+// fieldEncoder is no longer needed.
+func newFieldEncoder() *fieldEncoder {
+	fe := encPool.Get().(*fieldEncoder)
+	fe.fields = fe.fields[:0]
+	fe.namespace = fe.namespace[:0]
+	return fe
+}
+
+func (fe *fieldEncoder) Close() error {
+	encPool.Put(fe)
+	return nil
+}
+
+func (fe *fieldEncoder) AddArray(key string, marshaler zapcore.ArrayMarshaler) error {
+	// TODO(rfratto): allow this to write the value of the array instead of
+	// placeholder text.
+	fe.fields = append(fe.fields, fe.keyName(key), "<array>")
+	return nil
+}
+
+func (fe *fieldEncoder) AddObject(key string, marshaler zapcore.ObjectMarshaler) error {
+	// TODO(rfratto): allow this to write the value of the object instead of
+	// placeholder text.
+	fe.fields = append(fe.fields, fe.keyName(key), "<object>")
+	return nil
+}
```

Comment:
I didn't implement these here because:

1. I think they're going to be a massive pain.
2. I suspect logging arrays and objects is going to be fairly uncommon.